### PR TITLE
DEVOPS-2015 Redo pipeline and rename go package

### PR DESCRIPTION
**Pull request by richerve**
One open question that I have is if it's better to run the build process as part of the pipeline or as is done now as part of the Dockerfile build, that is run in the publish step. Opinions?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Drone plugin for Helm 3
 
 [![Build Status](https://cloud.drone.io/api/badges/pelotech/drone-helm3/status.svg)](https://cloud.drone.io/pelotech/drone-helm3)
```
Comment (dskatz) [inline README.md]: Build status should be removed, or updated to our drone.  Not sure what the status of allowing build badges is at.

**Review by dskatz — COMMENTED** [review]
There are still several parts in the README that have not updated, including the examples and badges.

File: README.md (hunk 1 above)
Comment (richerve) [reply README.md]: will do, thx. I think badges are allowed now, as far as I remember

Comment (richerve) [reply README.md]: updated the badge and adjusted other parts of Readme

File: .drone.yml [hunk 2]
```diff
@@ -1,32 +1,55 @@
...
+
   - name: lint
     image: cytopia/golint
     commands:
-      - golint -set_exit_status ./cmd/... ./internal/...
-  - name: build
-    image: golang:1.13
-    commands:
-      - go build -o build/drone-helm cmd/drone-helm/main.go
+      - golint -set_exit_status ./...
```
Comment (colinhoglund) [inline .drone.yml]: Let's combine these and use golangci-lint as we have started doing in other repos.

```
  - name: test
    image: golangci/golangci-lint:v1.30.0-alpine
    commands:
      - apk update
      - apk add make
      - make test
```

I believe this version uses Go 1.15, so you'll want to adjust accordingly.

File: .drone.yml [hunk 3]
```diff
@@ -1,32 +1,55 @@
...
+
   - name: publish_linux_amd64
     image: plugins/docker
     settings:
-      auto_tag: true
+      registry: quay.io
+      repo: quay.io/mongodb/drone-helm
```
Comment (colinhoglund) [inline .drone.yml:29]: Be sure to disable the drone repo for http://github.com/mongodb-forks/drone-helm

File: .drone.yml [hunk 4]
```diff
@@ -1,32 +1,55 @@
...
+      tags:
+        - git-${DRONE_COMMIT_SHA:0:7}
+        - ${DRONE_TAG}
+        - latest
+    when:
+      event:
+        - tag
+---
+kind: signature
+hmac: 8329e42fe25499744eb7522048352e62e7bf776fc0db251d549c4ef2826c790d
```
Comment (colinhoglund) [inline .drone.yml]: Any reason we're adding a signature here? We shouldn't be exposing any secrets to `pull_request` events.

File: README.md [hunk 5]
```diff
@@ -1,16 +1,18 @@
 # Drone plugin for Helm 3
 
-[![Build Status](https://cloud.drone.io/api/badges/pelotech/drone-helm3/status.svg)](https://cloud.drone.io/pelotech/drone-helm3)
-[![Go Report](https://goreportcard.com/badge/github.com/pelotech/drone-helm3)](https://goreportcard.com/report/github.com/pelotech/drone-helm3)
-[![](https://images.microbadger.com/badges/image/pelotech/drone-helm3.svg)](https://microbadger.com/images/pelotech/drone-helm3 "Get your own image badge on microbadger.com")
+[![Build Status](https://drone.corp.mongodb.com/api/badges/mongodb-forks/drone-helm3/status.svg)](https://drone.corp.mongodb.com/mongodb-forks/drone-helm3)
```
Comment (colinhoglund) [inline README.md:3]: you should remove this, build status doesn't work outside of our network.

File: .drone.yml (hunk 4 above)
Comment (richerve) [reply .drone.yml]: if this is not added then drone won't run when the pipeline is protected, or maybe I'm not understanding the usage for the signature

File: README.md (hunk 5 above)
Comment (richerve) [reply README.md:3]: /api/badges is now exposed publicly, but I can remove it if required

Comment (colinhoglund) [reply README.md:3]: oh, nm

File: .drone.yml (hunk 4 above)
Comment (colinhoglund) [reply .drone.yml]: i guess leaving it protected is a good idea since its a public repo. disregard.

File: .drone.yml (hunk 2 above)
Comment (richerve) [reply .drone.yml]: the pipeline is now using the golangci image

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 6]
```diff
@@ -1,20 +0,0 @@
----
```
Comment (colinhoglund) [inline .github/ISSUE_TEMPLATE/bug_report.md]: I don't see any reason to delete these templates.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,35 @@
+MODULE := github.com/mongodb-forks/drone-helm3
+CMD_NAME ?= drone-helm
+VERSION := $(shell grep '^const Version =' internal/version/version.go | cut -d\" -f2)
```
Comment (dskatz) [inline Makefile]: This is not used in the Makefile

File: internal/version/version.go [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package version
+
+// Version is the current version
+const Version = "1.0.0"
```
Comment (dskatz) [inline internal/version/version.go]: Should this start at 1.0?

Version also does not seem to be used anywhere in the application.  Is this file needed?

One place it could be used is to log the version to help with debugging.

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+RUN ?= .*
+PKG ?= ./...
+.PHONY: test
+test: ## Run tests in local environment
+	golangci-lint run --timeout=5m $(PKG)
+	go test -cover -run=$(RUN) $(PKG)
+
+.PHONY: docker
+docker: ## Build local development docker image with cached go modules, builds, and tests
+	@docker build -f build/Dockerfile-test -t $(CMD_NAME):latest .
```
Comment (colinhoglund) [inline Makefile:13]:
```suggestion
	@docker build -f Dockerfile -t $(CMD_NAME):latest .
```

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+docker-test: ## Run tests using local development docker image
+	@docker run -v $(shell pwd):/go/src/$(MODULE):delegated $(CMD_NAME) make test RUN=$(RUN) PKG=$(PKG)
+
+.PHONY: docker-snyk
+docker-snyk: ## Run local snyk scan, SNYK_TOKEN environment variable must be set
+	@docker run --rm -e SNYK_TOKEN -w /go/src/$(MODULE) -v $(shell pwd):/go/src/$(MODULE):delegated snyk/snyk:golang
+
+# build local docker image
+.PHONY: build
+build:
```
Comment (colinhoglund) [inline Makefile]: This is redundant with the `docker` target.

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	golangci-lint run --timeout=5m $(PKG)
+	go test -cover -run=$(RUN) $(PKG)
+
+.PHONY: docker
+docker: ## Build local development docker image with cached go modules, builds, and tests
+	@docker build -f build/Dockerfile-test -t $(CMD_NAME):latest .
+
+.PHONY: docker-test
+docker-test: ## Run tests using local development docker image
+	@docker run -v $(shell pwd):/go/src/$(MODULE):delegated $(CMD_NAME) make test RUN=$(RUN) PKG=$(PKG)
```
Comment (colinhoglund) [inline Makefile:17]: This command doesn't work when run with the provided `Dockerfile`.

File: internal/version/version.go (hunk 8 above)
Comment (colinhoglund) [reply internal/version/version.go]: +1, not really seeing the need from this package. Also, it seems like were on version `0.15.0` not v1

Comment (richerve) [reply internal/version/version.go]: ok, will remove it from the moment. Also the versioning will be confusing as it will produce an artifact tagged as v3, so I presume the software should start at that version really.

File: Makefile (hunk 11 above)
Comment (richerve) [reply Makefile:17]: It ran using the Dockerfile-test file

File: Makefile (hunk 9 above)
Comment (richerve) [reply Makefile:13]: the Dockerfile-test is for local development purposes and the Dockerfile is to produce the final artifact.

File: Makefile (hunk 10 above)
Comment (richerve) [reply Makefile]: the Dockerfile-test is for local development purposes and the Dockerfile is to produce the final artifact.

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 6 above)
Comment (richerve) [reply .github/ISSUE_TEMPLATE/bug_report.md]: reverted

File: Makefile (hunk 7 above)
Comment (richerve) [reply Makefile]: removed

File: internal/version/version.go (hunk 8 above)
Comment (richerve) [reply internal/version/version.go]: done

File: Makefile (hunk 9 above)
Comment (colinhoglund) [reply Makefile:13]: but there is no `build/Dockerfile-test`

Comment (richerve) [reply Makefile:13]: sorry, didn't notice that ./build was in gitignore, so wasn't pushing my changes there.

I see your point that "build" is probably redundant in terms of development as that's only required when publishing and that will happen in Drone.

File: .drone.yml [hunk 12]
```diff
@@ -1,32 +1,63 @@
...
-      auto_tag: true
+      registry: quay.io
+      repo: quay.io/mongodb/drone-helm
+      username:
+        from_secret: docker_username
+      password:
+        from_secret: docker_password
+      tags:
+        - git-${DRONE_COMMIT_SHA:0:7}
+        - latest
```
Comment (colinhoglund) [inline .drone.yml]: We shouldn't publish `latest` so that users are forced to pick a version.

File: .drone.yml [hunk 13]
```diff
@@ -1,32 +1,63 @@
...
         from_secret: docker_username
       password:
         from_secret: docker_password
-      repo:
-        from_secret: plugin_repo
-      dockerfile: Dockerfile
+      tags:
+        - git-${DRONE_COMMIT_SHA:0:7}
+        - ${DRONE_TAG}
+        - latest
```
Comment (colinhoglund) [inline .drone.yml]: We shouldn't publish `latest` so that users are forced to pick a version.

File: .drone.yml (hunk 12 above)
Comment (richerve) [reply .drone.yml]: done

File: .drone.yml (hunk 13 above)
Comment (richerve) [reply .drone.yml]: done

**Review by colinhoglund — APPROVED** [review]

**Review by colinhoglund — APPROVED** [review]
